services/st_util/StUartBootloaderCommandHandler.cpp

```diff
@@ -0,0 +1,552 @@
+        TryHandleTransmitAction();
+    }
+
+    void StUartBootloaderCommandHandler::TryHandleTransmitAction()
+    {
+        if (commandActions.front()->GetActionType() == ActionType::transmit)
+        {
+            static auto transmitaction = commandActions.front();
+            transmitaction = commandActions.front();
+
+            commandActions.pop_front();
+            transmitaction->SendData();
+        }
+    }
+
+    void StUartBootloaderCommandHandler::TryHandleNextAction()
+    {
+        commandActions.pop_front();
+
+        if (commandActions.empty())
+        {
+            timeout.Cancel();
+            onCommandExecuted();
+            return;
+        }
+
+        if (commandActions.front()->GetActionType() == ActionType::receive)
+            commandActions.front()->DataReceived();
+        else
+            TryHandleTransmitAction();
+    }
+
+    void StUartBootloaderCommandHandler::SendData(infra::ConstByteRange data, uint8_t checksum)
+    {
+        serial.SendData(data, [this, checksum]()
+            {
+                serial.SendData(infra::MakeConstByteRange(checksum), [this]()
+                    {
+                        TryHandleTransmitAction();
+                    });
+            });
+    }
+
+    void StUartBootloaderCommandHandler::SendData(infra::ConstByteRange data)
+    {
+        serial.SendData(data, [this]()
+            {
+                TryHandleTransmitAction();
+            });
+    }
+
+    void StUartBootloaderCommandHandler::DataReceived()
+    {
+        if (!commandActions.empty())
+            commandActions.front()->DataReceived();
+    }
+
+    void StUartBootloaderCommandHandler::OnError(infra::BoundedConstString reason)
+    {
+        timeout.Cancel();
+        serial.ReceiveData([](auto data) {});
+        if (onError)
+            onError(reason);
+    }
+
+    StUartBootloaderCommandHandler::Action::Action(StUartBootloaderCommandHandler& handler, StUartBootloaderCommandHandler::ActionType type)
+        : handler(handler)
+        , type(type)
+    {}
+
+    void StUartBootloaderCommandHandler::Action::SendData()
+    {
+        std::abort();
+    }
+
+    void StUartBootloaderCommandHandler::Action::DataReceived()
+    {
+        std::abort();
+    }
+
+    StUartBootloaderCommandHandler::ActionType StUartBootloaderCommandHandler::Action::GetActionType() const
+    {
+        return type;
+    }
+
+    StUartBootloaderCommandHandler::ReceiveAction::ReceiveAction(StUartBootloaderCommandHandler& handler)
+        : StUartBootloaderCommandHandler::Action(handler, StUartBootloaderCommandHandler::ActionType::receive)
+    {}
+
+    StUartBootloaderCommandHandler::TransmitAction::TransmitAction(StUartBootloaderCommandHandler& handler)
+        : StUartBootloaderCommandHandler::Action(handler, StUartBootloaderCommandHandler::ActionType::transmit)
+    {}
+
+    void StUartBootloaderCommandHandler::ReceiveAckAction::DataReceived()
+    {
+        auto byte = handler.queue.Get();
```

Comment:
In TryHandleNextAction, you invoke DataReceived to process any data received until then. However, it is not guaranteed that there is already data available. Therefore, handler.queue.Get() may fail, so this function needs to first check if there's already data available.